### PR TITLE
[Donated Buffer] support metadata mutation ops

### PR DESCRIPTION
**Pull request by BoyuanFeng**
### Background: 

`set(x,y)` changes the untyped storage of x to be the same as y.

```python
import torch
from torch._subclasses.fake_tensor import FakeTensorMode

x1 = torch.ones(2,3)
y1 = torch.ones(2,3)
z1 = torch.ops.aten.set_.source_Tensor(x1, y1)

fake_tensor_mode = FakeTensorMode()
x2 = fake_tensor_mode.from_tensor(torch.ones(2,3))
y2 = fake_tensor_mode.from_tensor(torch.ones(2,3))
z2 = torch.ops.aten.set_.source_Tensor(x2, y2)

print(f"x1: {x1.untyped_storage()._cdata}, y1: {y1.untyped_storage()._cdata}, z1: {z1.untyped_storage()._cdata}")
print(f"x2: {x2.untyped_storage()._cdata}, y2: {y2.untyped_storage()._cdata}, z2: {z2.untyped_storage()._cdata}")
# x1: 99973024, y1: 99973024, z1: 99973024
# x2: 112107232, y2: 112107232, z2: 112107232
```

### Error before this diff

Consider this example:
```python
import torch

def fn(x):
    p = torch.nn.Parameter(x + 123)
    return p, p.sin()

opt = torch.compile(fn, fullgraph=True)
x = torch.ones(16, device="cuda", requires_grad=True)

p, r = opt(x)
r.sum().backward()
```

When running with `TORCH_LOGS=aot`, we have `set_` in the graph.
```
def forward(self, primals_1: "f32[16][1]cuda:0", primals_2: "f32[16][1]cuda:0"):
   # File: /home/boyuan/playground/inductor/donated_buffer.py:4 in fn, code: p = torch.nn.Parameter(x + 123)
  add: "f32[16][1]cuda:0" = torch.ops.aten.add.Tensor(primals_1, 123);  primals_1 = None
  
   # File: /home/boyuan/playground/inductor/donated_buffer.py:5 in fn, code: return p, p.sin()
  sin: "f32[16][1]cuda:0" = torch.ops.aten.sin.default(add)
  
  # No stacktrace found for following nodes
  set_: "f32[16][1]cuda:0" = torch.ops.aten.set_.source_Tensor(primals_2, add);  primals_2 = set_ = None
  return (sin, add)
```

`set_: "f32[16][1]cuda:0" = torch.ops.aten.set_.source_Tensor(primals_2, add)` should change the storage of `primals_2` to be the same as `add`. However, this is not true before this diff. We found different untyped_storage() for meta['val'] of `set_`, `add`, and `primals_2`.

This also leads to an error with donated buffer (#130580), which checks alias by untyped_storage. Since `add` and `primals_2` have different untyped_storage (which is wrong), add is wrongly marked as donated buffer.

### Root Cause

During tracing, we have args, kwargs, out, and proxy_args, proxy_kwargs, proxy_out.

We use args and kwargs to compute `out = func(*args, **kwargs)` ([Here](https://github.com/pytorch/pytorch/blob/main/torch/fx/experimental/proxy_tensor.py#L912)). Later, we set out to its proxy, essentially calling `proxy_out.node.meta["val"] = out.detach()`.

Due to the detach, the storage change happens on args but not on proxy_args.node.meta["val"] when func is torch.ops.aten.set_. I repro'ed this behavior of detach in eager code.

```python
import torch

x = torch.ones(2,3)
x_detach = x.detach()
y = torch.ones(2,3)
z = torch.ops.aten.set_.source_Tensor(x_detach, y)


print(f"x: {x.untyped_storage()._cdata}, x_detach: {x_detach.untyped_storage()._cdata}, y: {y.untyped_storage()._cdata}, z: {z.untyped_storage()._cdata}")
# x: 97023632, x_detach: 97026480, y: 97026480, z: 97026480
```

To fix the issue, this PR manually resets node.meta["val"] if the storage has changed.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by ezyang** [comment]
@bdhirsh @yf225 i'm forcing you two to review this lol

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -911,6 +911,16 @@ def can_handle_tensor(x: Tensor) -> bool:
     with _enable_thunkify(proxy_mode.tracer):
         out = func(*args, **kwargs)
 
+    # reset meta val if func (e.g., set_) changed the untyped_storage of args
```
Comment (bdhirsh) [inline torch/fx/experimental/proxy_tensor.py:914]: Hmm so this feels like the wrong place to make the fix: in particular, we should be running the meta kernel for `aten.set_`, which should have properly swapped out the storage of the input for you already. So we'd have to dig into why that isn't happening inside of `FakeTensorMode`.

But @yf225 chatted ([comment](https://github.com/pytorch/pytorch/pull/140113#discussion_r1854663438)), and it sounds like:

(1) we don't actually need to maintain the sacrificial parameter code for FSDP2 tracing (We can rely on running the first iteration in eager, when the params are set up).

(2) FSDP2 tracing was the initial use case for adding the "sacrificial parameter logic. So it should be ok for us to start graph breaking on calls to `torch.nn.Parameter()`, which would be a nice simplification.

@BoyuanFeng how do you feel about updating dynamo to graph break on `nn.Parameter` instead? (we can kill the sacrificial parameter code in that PR as well. That would basically involve changing the code in dynamo that `traceable_create_parameter()` to raise `unimplemented("Parameter() on non-input")`. Basically a revert of these lines: https://github.com/pytorch/pytorch/pull/120965/files#diff-7bd43c6b174845f3450be42c8460ed021d1b39a98510cbea5d4e7e3a0b9d8d4fR844

Comment (bdhirsh) [reply torch/fx/experimental/proxy_tensor.py:914]: update: I misread. I agree with the problem, which is that the input node has a `meta['val']` field that needs to have its storage mutated, which proxy tensor needs to deal with.

This is probably another argument for banning metadata mutations from proxy tracing (they are painful), but doing it here sounds reasonable, and we can take "removing the sacrificial parameter logic" as a followup

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -911,6 +911,16 @@ def can_handle_tensor(x: Tensor) -> bool:
...
+    for tensor, proxy_tensor in zip(flat_args_kwargs, proxy_flat_args_kwargs):
+        if (
+            isinstance(tensor, FakeTensor)
+            and isinstance(proxy_tensor, Proxy)
+            and tensor.untyped_storage()
```
Comment (bdhirsh) [inline torch/fx/experimental/proxy_tensor.py:919]: This implementation requires looping over all of the args on extra proxy tensor call, and comparing each arguments' storage, even though we will perform the swap very rarely (metadata mutations are very rare). So this will have some compile time cost.

Instead, can you do something like this?
```
if torch.Tag.inplace_view in func.tags:
    # convention: all inplace_view ops mutate the metadata of their first argument
    arg = flat_args_kwargs[0]
    proxy_arg = proxy_flat_args_kwargs[0]
    assert isinstance(arg, torch.Tensor), f"invalid inplace_view op: {str(func)}"
    set_meta(proxy_arg, arg)
```

**Comment by ezyang** [comment]
The root cause of the problem is the "val" concept is incoherent in the presence of metadata mutation. We have attached a val to a node on an FX graph. Is this val the state of the tensor in question at the time of that node, or the "final" state of the tensor after the rest of the graph had run? I think the former is a lot more logical. But this means I have to snapshot nodes when I put them in val so they don't get mutated.

You seem to want some sort of ongoing evolving state of the fake tensors. We have that too, it's just not stored on val on the FX node, which is the wrong place to put it.

**Comment by ezyang** [comment]
So, what's the plan now?

**Review by bdhirsh — APPROVED** [review]
sgtm!
